package-recipe.el

```diff
@@ -60,6 +60,14 @@
       (format (oref rcp url-format)
               (oref rcp repo))))
 
+(cl-defmethod package-recipe--upstream-protocol ((rcp package-recipe))
+  (let ((url (package-recipe--upstream-url rcp)))
+    (if
+        (string-match "^\\([a-z]+\\):/" url)
```

Comment:
If you `(require 'url-parse)` you can do something like:
```el
(url-type (url-generic-parse-url "https://google.com"))
=> "https"
```

Reply:
Thanks for the advice! I hope it's better now!

Reply:
This doesn't work for one of the most common formats. Example: `git@github.com:foo/bar.git`.

It might be better to go back to what you had before and then adjust it to deal with this format as well.

Reply:
Oh, good point, sorry for the diversion.

Reply:
Thanks for the feedback!

There are now checks for the three main possible syntaxes (`protocol://`, scp-like, local file path).

Looking at the [git clone manpage](https://manpages.debian.org/buster/git-man/git-clone.1.en.html#GIT_URLS) there don't seem to be any other possibilities.  I believe `hg` only supports [the first and third syntaxes](https://www.mercurial-scm.org/repo/hg/help/urls) (explicit protocol or local path) and no others.

I've made the regex for the scp-like syntax highly permisive, to allow for some of the various possibilities:

1. `user@domain:path`

2. `domain:path`

3. `alias:path` (where the "alias" is defined in the `ssh_config`)

The regex could almost certainly be slightly more restrictive, but given that domain names can be internationalised or unqualified and that I'm not actually sure what restrictions are there for the host pattern in `ssh_config` (the man page states zero or more "non-whitespace characters" or one of the two wildcards), probably not significantly more so.  Also, it seems that an scp-like URL can't in general be distinguished from a local file path, without directory scanning (which would probably be an over-complication here), since an scp-like URL can also be a valid relative file path, and `git` will "prefer" the local file path, if it exists.  For instance, the following:

```bash
mkdir -p git@github.com:melpa/melpa
cd git@github.com:melpa/melpa
git init
cd -
git clone git@github.com:melpa/melpa
```

will cause the newly-created, empty repository to be cloned.